nixos/modules/hardware/video/nvidia.nix

```diff
@@ -18,6 +18,7 @@ let
   busIDType = lib.types.strMatching "([[:print:]]+[\:\@][0-9]{1,3}\:[0-9]{1,2}\:[0-9])?";
   ibtSupport = cfg.open || (nvidia_x11.ibtSupport or false);
   settingsFormat = pkgs.formats.keyValue { };
+  isOpen = cfg.open == true;
```

Comment:
Isn't the assertion enough to make sure cfg.open is a boolean?

Reply:
(Might need to reorder the assertions to change the evaluation order)

Reply:
I don't know, is there a way to validate that?

Reply:
I tried, well it didn't work. But this do work:

```diff
diff --git a/nixos/modules/hardware/video/nvidia.nix b/nixos/modules/hardware/video/nvidia.nix
index c6a2fa729c6f..39e2c1d79c43 100644
--- a/nixos/modules/hardware/video/nvidia.nix
+++ b/nixos/modules/hardware/video/nvidia.nix
@@ -18,7 +18,6 @@ let
   busIDType = lib.types.strMatching "([[:print:]]+[\:\@][0-9]{1,3}\:[0-9]{1,2}\:[0-9])?";
   ibtSupport = cfg.open || (nvidia_x11.ibtSupport or false);
   settingsFormat = pkgs.formats.keyValue { };
-  isOpen = cfg.open == true;
 in
 {
   options = {
@@ -256,12 +255,11 @@ in
       };
 
       open = lib.mkOption {
-        default = null;
         example = true;
         description = "Whether to enable the open source NVIDIA kernel module.";
-        type = lib.types.nullOr lib.types.bool;
+        type = lib.types.bool;
         defaultText = lib.literalExpression ''
-          if lib.versionOlder config.hardware.nvidia.package.version "560" then false else null
+          lib.mkIf (lib.versionOlder nvidia_x11.version "560") (lib.mkDefault false);
         '';
       };
 
@@ -322,7 +320,7 @@ in
           environment.systemPackages = [ nvidia_x11.bin ];
 
           hardware.nvidia.open = lib.mkIf (lib.versionOlder nvidia_x11.version "560") (lib.mkDefault false);
-          hardware.nvidia.gsp.enable = lib.mkDefault (isOpen || lib.versionAtLeast nvidia_x11.version "555");
+          hardware.nvidia.gsp.enable = lib.mkDefault (cfg.open || lib.versionAtLeast nvidia_x11.version "555");
         })
 
         # X11
@@ -386,20 +384,15 @@ in
             }
 
             {
-              assertion = isOpen -> (cfg.package ? open);
+              assertion = cfg.open -> (cfg.package ? open);
               message = "This version of NVIDIA driver does not provide a corresponding opensource kernel driver.";
             }
 
             {
-              assertion = isOpen -> cfg.gsp.enable;
+              assertion = cfg.open -> cfg.gsp.enable;
               message = "The GSP cannot be disabled when using the opensource kernel driver.";
             }
 
-            {
-              assertion = cfg.open != null;
-              message = "You must specify which variant of the nvidia kernel driver to use.";
-            }
-
             {
               assertion = cfg.dynamicBoost.enable -> lib.versionAtLeast nvidia_x11.version "510.39.01";
               message = "NVIDIA's Dynamic Boost feature only exists on versions >= 510.39.01";
@@ -599,7 +592,7 @@ in
               "L+ /run/nvidia-docker/extras/bin/nvidia-persistenced - - - - ${nvidia_x11.persistenced}/origBin/nvidia-persistenced";
 
           boot = {
-            extraModulePackages = if isOpen then [ nvidia_x11.open ] else [ nvidia_x11.bin ];
+            extraModulePackages = if cfg.open then [ nvidia_x11.open ] else [ nvidia_x11.bin ];
             # nvidia-uvm is required by CUDA applications.
             kernelModules =
               lib.optionals config.services.xserver.enable [
@@ -610,14 +603,14 @@ in
               # With the open driver, nvidia-uvm does not automatically load as
               # a softdep of the nvidia module, so we explicitly load it for now.
               # See https://github.com/NixOS/nixpkgs/issues/334180
-              ++ lib.optionals (config.services.xserver.enable && isOpen) [ "nvidia_uvm" ];
+              ++ lib.optionals (config.services.xserver.enable && cfg.open) [ "nvidia_uvm" ];
 
             # If requested enable modesetting via kernel parameters.
             kernelParams =
               lib.optional (offloadCfg.enable || cfg.modesetting.enable) "nvidia-drm.modeset=1"
               ++ lib.optional ((offloadCfg.enable || cfg.modesetting.enable) && lib.versionAtLeast nvidia_x11.version "545") "nvidia-drm.fbdev=1"
               ++ lib.optional cfg.powerManagement.enable "nvidia.NVreg_PreserveVideoMemoryAllocations=1"
-              ++ lib.optional isOpen "nvidia.NVreg_OpenRmEnableUnsupportedGpus=1"
+              ++ lib.optional cfg.open "nvidia.NVreg_OpenRmEnableUnsupportedGpus=1"
               ++ lib.optional (config.boot.kernelPackages.kernel.kernelAtLeast "6.2" && !ibtSupport) "ibt=off";
 
             # enable finegrained power management
```

Reply:
That just reversed the changes of this PR. The intention was to use `null` to signal the user had to choose.

Reply:
By marking the type `type = lib.types.bool;` and not providing `default`, user still has to choose.

Reply:
TIL, I'll apply the changes when I get the opportunity